### PR TITLE
Added resolution picker on consumption and production pages

### PR DESCRIPTION
**Pull request by NielsJorck**
## Description

Added resolution picker to be used in conjunction with date picker and graph data.
Can be shown on the page using ?enableFeature=resolution in the URL

image

## References

https://github.com/Energinet-DataHub/energy-origin-issues/issues/725

### CODE REVIEW

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+      }
+
+      mat-button-toggle {
+        border-color: var(--watt-color-primary) !important;
+      }
+
+      mat-button-toggle.mat-button-toggle-appearance-standard
+        .mat-button-toggle-label-content {
+        line-height: unset;
+        padding: 10px 14px;
```
Comment (MadsThy) [inline libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts:48]: Are these magical UX numbers? 😄

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+} from 'date-fns';
+import { map } from 'rxjs';
+
+@Component({
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  encapsulation: ViewEncapsulation.None,
+  selector: 'eo-resolution-picker',
+  styles: [
+    `
+      eo-resolution-picker {
```
Comment (MadsThy) [inline libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts:44]:
```suggestion
      :host {
```

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    `,
+  ],
+  template: `<label class="watt-space-stack-s title">Resolution</label>
+    <mat-button-toggle-group
+      [value]="appSettingsResolution$ | async"
+      name="resolution"
+    >
+      <mat-button-toggle
+        [disabled]="
+          !(((appSettingsCalenderDateRangeInDays$ | async) ?? 0) <= 7)
```
Comment (MadsThy) [inline libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts:84]: Did UX want us to make this appear as an option if we were under 3 days, or am i remember it wrong? :)

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        (change)="setResolution('HOUR')"
+        value="HOUR"
+        >Per Hour</mat-button-toggle
+      >
+      <mat-button-toggle (change)="setResolution('DAY')" value="DAY"
+        >Day</mat-button-toggle
+      >
+      <mat-button-toggle
+        [disabled]="
+          !(((appSettingsCalenderDateRangeInDays$ | async) ?? 0) >= 7)
```
Comment (MadsThy) [inline libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts:95]: I am really not in love with these negation definitions of why it should be disabled :) Can we maybe talk about how we can solve this so it becomes a little more readable about what it is we want?

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      map((x) => differenceInDays(x.end, x.start))
+    );
+  appSettingsCalenderDateRangeInMonths$ =
+    this.appSettingsStore.calenderDateRangeDates$.pipe(
+      map((x) => differenceInMonths(x.end, x.start))
+    );
+  appSettingsCalenderDateRangeInYears$ =
+    this.appSettingsStore.calenderDateRangeDates$.pipe(
+      map((x) => differenceInYears(x.end, x.start))
+    );
```
Comment (MadsThy) [inline libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts:131]:
```suggestion
  differenceIn$ = this.appSettingsStore.calenderDateRangeDates$.pipe(
    map((dates) => ({
      days: differenceInDays(dates.end, dates.start),
      months: differenceInMonths(dates.end, dates.start),
      years: differenceInYears(dates.end, dates.start) ?? 0,
    }))
  );
```

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+          !(((appSettingsCalenderDateRangeInMonths$ | async) ?? 0) >= 1)
+        "
+        (change)="setResolution('MONTH')"
+        value="MONTH"
+        >Month</mat-button-toggle
+      >
+      <mat-button-toggle
+        [disabled]="
+          !(((appSettingsCalenderDateRangeInYears$ | async) ?? 0) >= 1)
+        "
```
Comment (MadsThy) [inline libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts:112]:
```suggestion
       [disabled]="!(differenceIn.years >= 1)"
```

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+          !(((appSettingsCalenderDateRangeInDays$ | async) ?? 0) >= 7)
+        "
+        (change)="setResolution('WEEK')"
+        value="WEEK"
+        >Week</mat-button-toggle
+      >
+      <mat-button-toggle
+        [disabled]="
+          !(((appSettingsCalenderDateRangeInMonths$ | async) ?? 0) >= 1)
+        "
```
Comment (MadsThy) [inline libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts:104]:
```suggestion
        [disabled]="!(differenceIn.months >= 1)"
```

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        value="HOUR"
+        >Per Hour</mat-button-toggle
+      >
+      <mat-button-toggle (change)="setResolution('DAY')" value="DAY"
+        >Day</mat-button-toggle
+      >
+      <mat-button-toggle
+        [disabled]="
+          !(((appSettingsCalenderDateRangeInDays$ | async) ?? 0) >= 7)
+        "
```
Comment (MadsThy) [inline libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts:96]:
```suggestion
        [disabled]="!(differenceIn.days >= 7)"
```

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+  ],
+  template: `<label class="watt-space-stack-s title">Resolution</label>
+    <mat-button-toggle-group
+      [value]="appSettingsResolution$ | async"
+      name="resolution"
+    >
+      <mat-button-toggle
+        [disabled]="
+          !(((appSettingsCalenderDateRangeInDays$ | async) ?? 0) <= 7)
+        "
```
Comment (MadsThy) [inline libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts:85]:
```suggestion
        [disabled]="!(differenceIn.days <= 7)"
```

And did UX say something about wanting this to be when it is below 3 days? Or did I dream this thing up myself? :)

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+      mat-button-toggle.mat-button-toggle-appearance-standard.mat-button-toggle-checked {
+        background-color: var(--watt-color-primary);
+        color: var(--watt-color-neutral-white);
+      }
+    `,
+  ],
+  template: `<label class="watt-space-stack-s title">Resolution</label>
+    <mat-button-toggle-group
+      [value]="appSettingsResolution$ | async"
+      name="resolution"
```
Comment (MadsThy) [inline libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts:80]:
```suggestion
      name="resolution"
      *rxLet="differenceIn$ as differenceIn"
```

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+  setResolution(resolution: Resolution) {
+    this.appSettingsStore.setResolution(resolution);
+    this.newResolution.emit(resolution);
+  }
+}
+
+@NgModule({
+  declarations: [EoResolutionPickerComponent],
+  imports: [CommonModule, MatButtonToggleModule],
```
Comment (MadsThy) [inline libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts:146]:
```suggestion
  imports: [CommonModule, MatButtonToggleModule, LetModule],
```

File: libs/eo/shared/services/src/app-settings/app-settings.store.ts [hunk 12]
```diff
@@ -36,19 +40,55 @@ export class AppSettingsStore extends ComponentStore<AppSettingsState> {
...
+          break;
+        case days < 730:
+          defaultResolution = 'MONTH';
+          break;
+        case days > 730:
+          defaultResolution = 'YEAR';
+          break;
+      }
+      return { ...state, calendarDateRange, resolution: defaultResolution };
+    }
```
Comment (MadsThy) [inline libs/eo/shared/services/src/app-settings/app-settings.store.ts:84]:
```suggestion
    (state, calendarDateRange: CalendarDateRange): AppSettingsState => {
      const difference = differenceInDays(
        new Date(calendarDateRange.end),
        new Date(calendarDateRange.start)
      );

      return {
        ...state,
        calendarDateRange,
        resolution: this.getResolutionFromDifference(difference),
      };
    }
```

File: libs/eo/shared/services/src/app-settings/app-settings.store.ts [hunk 13]
```diff
@@ -36,19 +40,55 @@ export class AppSettingsStore extends ComponentStore<AppSettingsState> {
...
+  );
+
+  readonly resolution$ = this.select((state) => state.resolution);
+  readonly setResolution = this.updater(
+    (state, resolution: Resolution): AppSettingsState => ({
       ...state,
-      calendarDateRange,
+      resolution,
     })
   );
```
Comment (MadsThy) [inline libs/eo/shared/services/src/app-settings/app-settings.store.ts:93]:
```suggestion
  );
  
  getResolutionFromDifference(differenceInDays: number): Resolution {
    switch (true) {
      case differenceInDays < 3:
        return 'HOUR';
      case differenceInDays < 30:
        return 'DAY';
      case differenceInDays < 180:
        return 'WEEK';
      case differenceInDays < 730:
        return 'MONTH';
      case differenceInDays > 730:
        return 'YEAR';
      default:
        return 'MONTH';
    }
  }
```

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts (hunk 1 above)
Comment (NielsJorck) [reply libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts]: Yes :)

File: .github/workflows/eo-cd.yml [hunk 14]
```diff
@@ -236,6 +236,7 @@ jobs:
           chart-folder: build/infrastructure/eo/chart
           chart-repository: ${{ github.repository_owner }}/helm-charts
           chart-version: ${{ needs.define_build.outputs.chart_version }}
+          version_yaml_path_prefix: ''
```
Comment (MadsThy) [inline .github/workflows/eo-cd.yml:239]: What is this? ❓

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+
+      mat-button-toggle.mat-button-toggle-appearance-standard {
+        border-color: var(--watt-color-primary);
+      }
+
+      mat-button-toggle.mat-button-toggle-appearance-standard
+        .mat-button-toggle-label-content {
+        line-height: unset;
+        /* Magic UX number */
+        padding: 9px 14px;
```
Comment (MadsThy) [inline libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-resolution-picker/eo-resolution-picker.component.ts:66]:
```suggestion
        padding: 9px 14px; /* Magic UX number */
```

File: .github/workflows/eo-cd.yml (hunk 14 above)
Comment (NielsJorck) [reply .github/workflows/eo-cd.yml:239]: Very important 😎
Fixes the pipeline for preview.

Comment (NielsJorck) [reply .github/workflows/eo-cd.yml:239]: Honestly should probably have been a completely seperate PR.

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-stack/eo-stack.component.ts [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+      }
+    `,
+  ],
+  template: `<ng-content></ng-content>`,
+})
+export class EoStackComponent {
+  @Input() size: StackSize = 'L';
+  @HostBinding('attr.data-size') get stackSize() {
+    return this.size.toLowerCase();
+  }
```
Comment (MadsThy) [inline libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-stack/eo-stack.component.ts:70]:
```suggestion
    @Input() size: StackSize = 'L';
```

File: .github/workflows/eo-cd.yml (hunk 14 above)
Comment (MadsThy) [reply .github/workflows/eo-cd.yml:239]: No no :) Its fine with me. Just happy to know why it's there.

File: libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-stack/eo-stack.component.ts [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  ChangeDetectionStrategy,
+  Component,
+  HostBinding,
```
Comment (MadsThy) [inline libs/eo/shared/atomic-design/ui-atoms/src/lib/eo-stack/eo-stack.component.ts:20]: can get yeeted out of here :)

**Review by MadsThy — APPROVED** [review]
Just the one thing remaining, then this is gold 😄

**Review by MadsThy — APPROVED** [review]

**Pull request closed without merging**